### ISSUE TITLE
RobertaForMaskedLM.from_pretrained throwing some weights not initialized error when loading same model type

### ISSUE THREAD

**Issue by dhuruvasaditya** [huggingface/transformers#12232]
## Environment info

- `transformers` version: 4.7.0
- Platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.10
- PyTorch version (GPU?): 1.8.1+cu101 (False)
- Tensorflow version (GPU?): 2.5.0 (False)
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

## Information

I am using a pretrained RobertaForMaskedLM . When I try to load the model I get the following error:

Some weights of RobertaForMaskedLM were not initialized from the model checkpoint at mldmm/GlassBERTa and are newly initialized: ['lm_head.layer_norm.bias', 'lm_head.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.weight', 'lm_head.decoder.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.

The config of the model is as follows:
```
{
  "architectures": [
    "RobertaForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": 0,
  "eos_token_id": 2,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dim": 96,
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "roberta",
  "n_heads": 3,
  "n_layers": 3,
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 1,
  "position_embedding_type": "absolute",
  "transformers_version": "4.6.1",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 541
}
```
## To reproduce

```
from transformers import AutoModelForMaskedLM,AutoTokenizer
tok = AutoTokenizer.from_pretrained('mldmm/GlassBERTa')
mod = AutoModelForMaskedLM.from_pretrained('mldmm/GlassBERTa')
```
or
```
from transformers import RobertaForMaskedLM,AutoTokenizer
tok = AutoTokenizer.from_pretrained('mldmm/GlassBERTa')
mod = RobertaForMaskedLM.from_pretrained('mldmm/GlassBERTa')
```

Same goes for many other pre-trained models hosted ('beomi/kcbert-base','hfl/chinese-bert-wwm-ext')
## Expected behavior
Since I'm loading for the same architecture I am expecting a clean import without any errors, might lead to random output due to those newly initialized layers and can mess the output in fill-mask pipeline.
